presto-main/src/main/java/com/facebook/presto/operator/aggregation/histogram/GroupedTypedHistogram.java

```diff
@@ -0,0 +1,548 @@
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static it.unimi.dsi.fastutil.HashCommon.arraySize;
+import static it.unimi.dsi.fastutil.HashCommon.murmurHash3;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * implementation that uses groupId in the hash key, so that we may store all groupId x value -> count in one giant map. The value, however,
+ * is normalized into a single ValueBuilder. Further, we do not have per-group instances of objects. In order to construct a histogram for a group on
+ * demand, a linked list for each group is stored that overlays all virtual nodes (a shared index, i, across parallel arrays).
+ * <p>
+ * <pre>
+ * eg,
+ * heads[100] -> 3
+ *
+ * means the first piece of data is (values[valuePositions[3]], counts[3])
+ * If next[3] is 10, then we look at (values[valuePositions[10]], counts[10])
+ *
+ * and so on to construct the value, count pairs needed. An iterator-style function, readAllValues, exists and the caller may do whatever want with the
+ * values.
+ *
+ * </pre>
+ */
+
+public class GroupedTypedHistogram
+        implements TypedHistogram
+{
+    static final float MAX_FILL_RATIO = 0.5f;
+
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(GroupedTypedHistogram.class).instanceSize();
+    private static final int EMPTY_BUCKET = -1;
+    private static final int NULL = -1;
+    // JMH work indicated this was superior to linear or sum_of_count
+    private static final ProbeType PROBE_TYPE = ProbeType.SUM_OF_SQUARE;
+    private final int bucketId;
+
+    private final Type type;
+    private final BlockBuilder values;
+    private final BucketNodeFactory bucketNodeFactory;
+    //** these parallel arrays represent a node in the hash table
+    private final LongBigArray counts;
+    // need to store the groupId for a node for when we are doing value comparisons in hash lookups
+    private final LongBigArray groupIds;
+    // array of nodePointers (index in counts, valuePositions)
+    private final IntBigArray nextPointers;
+    // since we store histogram values in a hash, two histograms may have the same position (each unique value should have only one position in the internal
+    // BlockBuilder of values; not that we extract a subset of this when constructing per-group-id histograms)
+    private final IntBigArray valuePositions;
+    private final LongBigArray valueHashList = new LongBigArray();
+    //** end per-node arrays
+
+    // per groupId, we have a pointer to the first in the list of nodes for this group
+    private final LongBigArray headPointers;
+
+    private IntBigArray buckets;
+    private int nextNodePointer;
+    private int mask;
+    private int bucketCount;
+    private int maxFill;
+    // at most one thread uses this object at one time, so this must be set to the group being operated on
+    private long currentGroupId = -1;
```

Comment:
hmm, but i thought the opposite was where we wanted to go: GroupTypedHistogram IS the state, but due to self-referential issues with making it the state (how can it deserialize itself without existing?), we choose to keep a lightweight state over it

the state is really a factory/wrapper. I think it would obfuscate things because GroupTypedHistogram stands on its own as-is. the only vars in the wrapper state are

1. the histogram
2. the size (which we could work with)


oh also, i tried this initially--i hit some roadblocks. maybe you could help me figure them out. The TypedHistogram is an interface. woudl we make the single always take 0? it meant changing the single interface to accommodate the grouped.

i actually remember not liking the mutable state either

let's talk and see

Reply:
which also meant more mods to generated code which, at the time, i had 0 understanding of. i have slightly > 0 now

Reply:
as we agreed, this is a good idea, but tricky and requires agg framework changes

this function is called directly w/o groupId; the groupId is "state" that the agg state has access to , and passes to this via a setter

i don't like it either